Chore/create image placeholder

JamesRobertSutcliffe opened the pull request:
- Code added to render logo as placeholder if user does not have profile picture - can you please test this as I believe you do not have a profile picture saved? 

Also some general cleanup with re-optimisation for typescript.

patdel0 commented on src/app/personalised-homepage/page.js:
```diff
@@ -83,10 +83,15 @@ function PersonalisedHomepage() {
+    // UserProfilePic function checks whether user 
+
+    const UserProfilePic = () => {
+        return userImage !== null
```
When the user doesn't have an image, `userImage` becomes an empty string. 

This means that this statement will always return true, regardless of the user having a profile picture or not.

```js
"" !== null // returns true
```

patdel0 reviewed (changes requested):
It's not working as expected. 
With the `!== null` condition, it doesn't display the image at all. 

When the condition is changed, the image appears for a couple of seconds, then disappears. This suggests to me that there is a state that is changing the value of `userImage`.

Happy to take over this one if you want. It would be a good opportunity to learn and create some tests for this one though.

JamesRobertSutcliffe commented:
Thanks for this, I will make the change to this and try and write some tests!

JamesRobertSutcliffe commented:
I have added some commits. Here is what I have done / tried to achieve - 

1- I have created a placeholder "Loading" component that appears until profile data is fetched. This stops the image from flipping. I intend to change the loading text to a slicker loading animation in the future
2 - I have changed the dependency on displaying the placeholder logo to on the data fetched being undefined - hopefully this fixes. 
3 - I have created tests in the 'profilePic.spec.ts' file - I was a little unsure what I was doing here or if I did it correctly so not sure how useful they are
4 - Please also test using your profile - hopefully the previous issues of logo displaying innappropriately / flicking between logo / profile pic is not fixed

I have also noticed that there is not a deployment issue. I do not have time to fix now. Let me know if you would prefer I fix this before checking, if not I can try and fix after you text / review.

patdel0 commented on tests/Frontend/profilePic.spec.ts:
```diff
@@ -0,0 +1,14 @@
+test("Users profile picture should display if present", () => {
+    const userImage = "https://i.scdn.co/image/ab6775700000ee85357ee0eabf98dc740c83bb9c"
+    const setProfilePic = userImage === undefined ? "https://i.ibb.co/WHfbS7L/logo.png" : userImage;
+    expect(setProfilePic).toBe(userImage);
```
Nice one on starting out with Jest and testing! 👍

**No Action required** 

Right now, the test just checks if JavaScript and its ternary operators are doing their thing. It's not really looking at the component we're trying to test.

To focus on the component, you might want to test the actual DOM elements:

```js
import { render, screen } from '@testing-library/react';
import PersonalisedHomepage from '../PersonalisedHomepage';

it('should show the profile image', () => {
  render(<PersonalisedHomepage />);

  const imgElement = screen.getByRole('img');

  expect(imgElement).toBeInTheDocument();
  expect(imgElement).toHaveAttribute('src', 'expected-image-url');
});
```
To make this work, you'll need to add a package like `@testing-library/react` for DOM testing with Jest.

Here we are testing the front end and what the component renders in the document.

patdel0 replied:
This code is for demonstration purposes only.

patdel0 reviewed (approved):
Nice job making it work! 

<img width="1421" alt="image" src="https://github.com/JamesRobertSutcliffe/personality-music-recommender/assets/53922624/aa7d0b2a-d656-4060-b006-710abdb2e15a">
